src/basic/stat-util.c

```diff
@@ -75,7 +75,9 @@ int dir_is_empty_at(int dir_fd, const char *path) {
         if (path)
                 fd = openat(dir_fd, path, O_RDONLY|O_DIRECTORY|O_CLOEXEC);
         else
-                fd = fcntl(fd, F_DUPFD_CLOEXEC, 3);
+                /* Note that DUPing is not enough, as the internal pointer
+                 * would still be shared and moved by FOREACH_DIRENT. */
+                fd = fd_reopen(dir_fd, O_CLOEXEC);
         if (fd < 0)
                 return -errno;
```

Comment:
fd_reopen() returns the error code in its return value while openat() returns it in errno. hence needs to be:

if (path) {
        fd = openat(…);
        if (fd < 0)
                return -errno;
} else {
        fd = fd_reopen(…);
        if (fd < 0)
                return fd;
}
…
```

Reply:
whops good catch, fixed - ptal